File: 2530-minimize-maximum-of-array/2530-minimize-maximum-of-array.py
class Solution:
    def minimizeArrayValue(self, nums: List[int]) -> int:
        currentMax = nums[0]
        excess = 0
        sumSoFar = nums[0]

        for i in range(1, len(nums)):
            sumSoFar += nums[i]
            
            if nums[i] - currentMax > excess:
                nums[i] -= excess
                currentMax += ceil((nums[i] - currentMax) / (i + 1))

            excess = currentMax * (i + 1) - sumSoFar

        return currentMax
                

File: 2530-minimize-maximum-of-array/2530-minimize-maximum-of-array.cpp
class Solution {
public:
    int minimizeArrayValue(vector<int>& nums) {
        long sum = 0;
        int ans = INT_MIN;

        for (int i = 0; i < nums.size(); i++) {
            int count = i + 1;
            sum += nums[i];
            ans = max(ans, int((sum + i) / (i + 1)));
        }

        return ans;
    }
};